File: apps/api/src/service/UserService.ts
// Handle business logic related to users

import UserRepository from '../repository/UserRepository';

class UserService {
    static async getAllUsers() {
        const users = await UserRepository.getAllUsers();
        return users;
    }

    static async getUserById(userId: string) {
        const user = await UserRepository.getUserById(userId);
        return user;
    }

    static async createUser(id: string, name: string, email: string, passwordHash: string) {
        const newUser = await UserRepository.createUser(id, name, email, passwordHash);
        return newUser;
    }

    static async updateUser(userId: string, name?: string, email?: string) {
        const updatedUser = await UserRepository.updateUser(userId, name, email);
        return updatedUser;
    }

    static async deleteUser(userId: string) {
        await UserRepository.deleteUser(userId);
    }
}

export default UserService;

File: apps/cs-api/Services/UserService.cs
using cs_api.Models;
using cs_api.Repositories;

namespace cs_api.Services;

public class UserService : IUserService
{
    private readonly IUserRepository _userRepository;

    public UserService(IUserRepository userRepository)
    {
        _userRepository = userRepository;
    }

    public async Task<IEnumerable<User>> GetAllUsers()
    {
        return await _userRepository.GetAllUsers();
    }

    public async Task<User?> GetUserById(string userId)
    {
        return await _userRepository.GetUserById(userId);
    }

    public async Task<User> CreateUser(string userName, string email, string password)
    {
        var userId = Guid.NewGuid().ToString();
        var passwordHash = BCrypt.Net.BCrypt.HashPassword(password, workFactor: 10);
        return await _userRepository.CreateUser(userId, userName, email, passwordHash);
    }

    public async Task<User?> UpdateUser(string userId, string? userName, string? email)
    {
        return await _userRepository.UpdateUser(userId, userName, email);
    }

    public async Task DeleteUser(string userId)
    {
        await _userRepository.DeleteUser(userId);
    }
}
